[WIP] Subscriptions instead of License keys

rooftopcellist opened the pull request:
##### SUMMARY

Overhaul of licensing in favor of using RH subscriptions model.  

![subscriptions-ui-flow](https://user-images.githubusercontent.com/11698892/93937342-9712cc00-fcf5-11ea-85a2-41ba9404568b.gif)



##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - UI
 - Installer

rooftopcellist committed:
init subscriptions validation

rooftopcellist committed:
init subscriptions validation

rooftopcellist committed:
Add Licensing class
 - parse entitlement cert for subscription metadata
 - remove StubLicense
 - remove unneeded license_key and references

rooftopcellist committed:
Add Licensing class
 - parse entitlement cert for subscription metadata
 - remove StubLicense
 - remove unneeded license_key and references

rooftopcellist committed:
Subscriptions: Consolidate cert/key to one setting
  * Accept cert/key pair via a POST to the config endpoint
  * Have the API `api/v2/config` endpoint handle validating and saving the cert.

rooftopcellist committed:
Subscriptions: Consolidate cert/key to one setting
  * Accept cert/key pair via a POST to the config endpoint
  * Have the API `api/v2/config` endpoint handle validating and saving the cert.

rooftopcellist committed:
Subscriptions: Complete backend for handling Entitlement cert
  * remove ability to generate license name
  * remove company_name from license metadata
  * put in dummy validator for entitlement cert
  * format entitlement cert/key and write it to temp file so that rct can parse it

rooftopcellist committed:
Subscriptions: Complete backend for handling Entitlement cert
  * remove ability to generate license name
  * remove company_name from license metadata
  * put in dummy validator for entitlement cert
  * format entitlement cert/key and write it to temp file so that rct can parse it

rooftopcellist committed:
Subscriptions: Correct License Metadata
  * return "UNLICENSED" metadata when no entitlement cert is provided for Tower
  * save AWX license metadata to LICENSE setting when on AWX
  * Add Product Name field to metadata
  * Change X-API-Product-Name to key off of settings.LICENSE for faster look-up

rooftopcellist committed:
Subscriptions: Correct License Metadata
  * return "UNLICENSED" metadata when no entitlement cert is provided for Tower
  * save AWX license metadata to LICENSE setting when on AWX
  * Add Product Name field to metadata
  * Change X-API-Product-Name to key off of settings.LICENSE for faster look-up

rooftopcellist committed:
Fix Product-Name header for production instances

rooftopcellist committed:
Fix Product-Name header for production instances

rooftopcellist committed:
Subscriptions: clear Entitlement Cert out of the database on config delete

rooftopcellist committed:
Subscriptions: clear Entitlement Cert out of the database on config delete

rooftopcellist committed:
Subscriptions: Remove unnecessary default on cert setting

rooftopcellist committed:
Subscriptions: Remove unnecessary default on cert setting

mabashian committed:
Updates license upload to work with entitlement/key pair instead of json license.  Updates strings to reference Subscription/Entitlement instead of License

mabashian committed:
Updates license upload to work with entitlement/key pair instead of json license.  Updates strings to reference Subscription/Entitlement instead of License

rooftopcellist committed:
Subscriptions: Initial pass at Verifying key with CDN

rooftopcellist committed:
Subscriptions: Initial pass at Verifying key with CDN

rooftopcellist committed:
Subscriptions: Validate cert tempfile with CDN

rooftopcellist committed:
Subscriptions: Validate cert tempfile with CDN

rooftopcellist committed:
Subscriptions: Add Attach endpoint to apply the chosen subscription
  - Bump analytics config data version as this feature changes license data

rooftopcellist committed:
Subscriptions: Add Attach endpoint to apply the chosen subscription
  - Bump analytics config data version as this feature changes license data

rooftopcellist committed:
connect to subscription dots, attach sub and get the cert

rooftopcellist committed:
connect to subscription dots, attach sub and get the cert

chrismeyersfsu committed:
add rhsm python dep

chrismeyersfsu committed:
add rhsm python dep

rooftopcellist committed:
Subscriptions: Fix redundant pool_ids

rooftopcellist committed:
Subscriptions: Fix redundant pool_ids

rooftopcellist committed:
Subscriptions: Complete Attach and Login flow
  - Add ENTITLEMENT_CONSUMER setting to surface consumer uuid to user
  - Create meta consumer for the entire cluster installation
  - Attach user chosen sub to meta consumer & store entitlement serial #
  - Use serial # to retrieve meta consumers entitlement certificate for the attached pool
  - Refactor logic for clearing license data when invalid into a function

rooftopcellist committed:
Subscriptions: Complete Attach and Login flow
  - Add ENTITLEMENT_CONSUMER setting to surface consumer uuid to user
  - Create meta consumer for the entire cluster installation
  - Attach user chosen sub to meta consumer & store entitlement serial #
  - Use serial # to retrieve meta consumers entitlement certificate for the attached pool
  - Refactor logic for clearing license data when invalid into a function

chrismeyersfsu committed:
add rhsm python dep from github

chrismeyersfsu committed:
add rhsm python dep from github

rooftopcellist committed:
Subscriptions: Always specify org when registering consumer
  - Request the user provide an org if there are more than one on the RH account

rooftopcellist committed:
Subscriptions: Always specify org when registering consumer
  - Request the user provide an org if there are more than one on the RH account

rooftopcellist committed:
Automatically create consumer if only one org on the RH account

rooftopcellist committed:
Automatically create consumer if only one org on the RH account

rooftopcellist committed:
Add exception handling for if the user provides an org that does not exist

rooftopcellist committed:
Add exception handling for if the user provides an org that does not exist

rooftopcellist committed:
Subscriptions: Add subscription specific username & pw settigns
  * If the user submits an org id that is not needed, raise an error
  * Save org id on the consumer setting
  * Rename payload from "rh_username" --> "subscriptions_username"

rooftopcellist committed:
Subscriptions: Add subscription specific username & pw settigns
  * If the user submits an org id that is not needed, raise an error
  * Save org id on the consumer setting
  * Rename payload from "rh_username" --> "subscriptions_username"

rooftopcellist committed:
Subscriptions: Add subman facts to meta-consumer to get correct cert
  * If the certificate_version fact is not passed, the pool_id is not
    included in the entitlement certificate

rooftopcellist committed:
Subscriptions: Add subman facts to meta-consumer to get correct cert
  * If the certificate_version fact is not passed, the pool_id is not
    included in the entitlement certificate

rooftopcellist committed:
Subscriptions: Update error message for subscriptions payload

rooftopcellist committed:
Subscriptions: Update error message for subscriptions payload

Commit:
update ui to support new subscriptions credentials

Commit:
update ui to support new subscriptions credentials

rooftopcellist committed:
Subscriptions: Parse datetime properly from entitlement cert

rooftopcellist committed:
Subscriptions: Parse datetime properly from entitlement cert

chrismeyersfsu committed:
add subscription cert refresh

* add refresh code that will, given a consumer_id from the
settings.ENTITLEMENT_CONSUMER['consumer_id'] field, will hit RHSM or
Satellite API to find the 1-and-only-1 entitlement that is associated,
will regenerate all certs related to that entitlement (should only be
one), pull down the new refreshed cert, and save it to the dabase in
settings.ENTITLEMENT_CERT
* add /api/v2/config/refresh endpoint that calls refresh code
* add refresh_subscription manage command

rooftopcellist committed:
Subscriptions: Update UI wording for Sat orgs
  * Clean up flake8

rooftopcellist committed:
Subscriptions: Update UI wording for Sat orgs
  * Clean up flake8

mabashian commented on awx/ui/client/src/license/license.controller.js:
```diff
@@ -44,11 +44,11 @@ export default
+                $scope.title = i18n._("Subscription Management");
             }
-
+            // What if I POST the file as a field in the License data JSON?
```
Still need this?

mabashian commented on awx/ui_next/src/variables.js:
```diff
@@ -1,2 +1,2 @@
-export const BrandName = 'AWX';
```
Probably don't want to commit this change to devel

jlmitch5 commented on awx/ui/client/src/license/license.partial.html:
```diff
@@ -79,7 +73,7 @@
 			<div class="List-titleText">{{title}}</div>
 			<div class="License-body">
-				<div class="License-helperText License-introText" ng-if="licenseMissing" translate>Welcome to Ansible Tower! Please complete the steps below to acquire a license.</div>
+				<div class="License-helperText License-introText" ng-if="licenseMissing" translate>Welcome to Ansible Tower! Please complete the steps below to acquire a Subscription.</div>
```
sentence casing

jlmitch5 commented on awx/ui/client/src/license/license.partial.html:
```diff
@@ -88,29 +82,29 @@
 								</span>
 								<span class="License-helperText">
-									<translate>Please click the button below to visit Ansible's website to get a Tower license key.</translate>
+									<translate>Please click the button below to visit Ansible's website to get a Tower Subscription.</translate>
```
sentence casing

jlmitch5 commented on awx/ui/client/src/license/license.partial.html:
```diff
@@ -125,12 +119,12 @@
 							<div class="AddPermissions-directions">
 								<span class="License-helperText">
-									<translate>Provide your Red Hat customer credentials and you can choose from a list of your available licenses.  The credentials you use will be stored for future use in retrieving renewal or expanded licenses. You can update or remove them in SETTINGS > SYSTEM.</translate>
+									<translate>Provide your Red Hat or Satellite credentials below and you can choose from a list of your available Subscriptions.  If you are a Satellite user please specify the organization ID you wish to associate this Tower Installation to.  The credentials you use will be stored for future use in retrieving renewal or expanded Subscriptions.</translate>
```
sentence casing

jlmitch5 reviewed (commented):
Subscription should not be capitalized in these places

jlmitch5 committed:
remove unnecessary comment

jlmitch5 committed:
remove unnecessary comment

jlmitch5 committed:
fix unnecessarily commited brand name change

jlmitch5 committed:
fix unnecessarily commited brand name change

rooftopcellist committed:
Fix casing for UI subscriptions page

rooftopcellist committed:
Fix casing for UI subscriptions page

rooftopcellist committed:
Subscriptions: Fix upload flow, not being able to find cert in payload

rooftopcellist committed:
Subscriptions: Fix upload flow, not being able to find cert in payload

jlmitch5 committed:
add back in eula acceptance for license file based add

jlmitch5 committed:
add back in eula acceptance for license file based add

rooftopcellist committed:
Subscriptions: fix 500 on eula POST to config

rooftopcellist committed:
Subscriptions: fix 500 on eula POST to config

rooftopcellist committed:
Subscriptions: Refactor to make GET requests to config faster

rooftopcellist committed:
Subscriptions: Refactor to make GET requests to config faster

rooftopcellist committed:
Add custom facts for linking with automation analytics

rooftopcellist committed:
Add custom facts for linking with automation analytics

rooftopcellist committed:
Subscriptions: Get content base_url from subman

rooftopcellist committed:
Subscriptions: Get content base_url from subman

chrismeyersfsu committed:
remove roll-up filtering in favor of other filter

* Remove roll-up filtering based on date
* Keep expired and validity filter
* Keep 480 product filter
* Remove filter that allows subscriptions based on Name

chrismeyersfsu committed:
remove roll-up filtering in favor of other filter

* Remove roll-up filtering based on date
* Keep expired and validity filter
* Keep 480 product filter
* Remove filter that allows subscriptions based on Name

elyezer committed:
Update awxkit subscriptions page object

rooftopcellist committed:
Subscriptions: add os import

rooftopcellist committed:
Subscriptions: add os import

rooftopcellist committed:
Subscriptions: Unattach subs when needed
  * Unattach sub if given pool_id is already attached to avoid 403
  * Get and format candlepin host from subman
  * Exception handling for missing consumer on sat/cloud

rooftopcellist committed:
Subscriptions: Unattach subs when needed
  * Unattach sub if given pool_id is already attached to avoid 403
  * Get and format candlepin host from subman
  * Exception handling for missing consumer on sat/cloud

rooftopcellist committed:
Subscriptions: Clear config when old license key is encountered

rooftopcellist committed:
Subscriptions: Clear config when old license key is encountered

rooftopcellist committed:
Subscriptions: Make end_date int and add error handling

rooftopcellist committed:
Subscriptions: Make end_date int and add error handling

rooftopcellist committed:
Subscriptions: fix caching issue with applied config

rooftopcellist committed:
Subscriptions: fix caching issue with applied config

rooftopcellist committed:
Merge pull request #10 from rooftopcellist/certs_unattach_old_subs

Subscriptions: Unattach subs when done

rooftopcellist committed:
Fix 500 by handling license not existing

rooftopcellist committed:
Fix 500 by handling license not existing

rooftopcellist committed:
Use RH content url and path settings

rooftopcellist committed:
Use RH content url and path settings

wenottingham committed:
Use rhsm libs for parsing certificate info

rooftopcellist committed:
Merge pull request #13 from shanemcd/certs

Certs

rooftopcellist committed:
Update entitlement_cert setting to fix bug with upload

rooftopcellist committed:
add valid_key and type to license data

shanemcd committed:
Allow for passing extra options to docker-compose up

rooftopcellist committed:
Merge pull request #15 from shanemcd/certs

Allow for passing extra options to docker-compose up

rooftopcellist committed:
Subscriptions: Fix missing SKU on login flow

shanemcd commented on awx/main/validators.py:
```diff
@@ -17,6 +17,12 @@
 
 
+def validate_entitlement_cert(data, min_keys=0, max_keys=None, min_certs=0, max_certs=None):
+    # TODO: replace with more actual robust logic here to allow for multiple certificates in one cert file (because this is how entitlements do)
```
This comment exists to make sure we dont forget about this.

shanemcd commented on requirements/requirements_git.txt:
```diff
@@ -1 +1,2 @@
 git+https://github.com/ansible/system-certifi.git@devel#egg=certifi
+git+https://github.com/candlepin/python-rhsm.git@python3-backport#egg=rhsm
```
We should fork this under the Ansible org.

shanemcd reviewed (changes requested):
Overall looks pretty good. Just added a couple small notes on things we dont want to forget.

rooftopcellist committed:
Subscriptions: Fix settings cache leaky reference issue
  * _attrs was using UNLICENSED_DATA multiple times, which
    used the same pointer/reference.  By using .copy, we now
    use a different pointer each time, which solves the problem.

rooftopcellist committed:
Subscriptions: Fix flake8 and add support_level to config data

ryanpetrello commented on awx/main/utils/licensing.py:
```diff
@@ -0,0 +1,356 @@
+            license_type='UNLICENSED',
+        )
+        if not kwargs:
+            license_setting = Setting.objects.filter(key='LICENSE').first()
```
Shouldn't this just be `settings.LICENSE` ?  Is this something you were experimenting with in relation to https://github.com/ansible/awx/pull/8209/commits/5ef2c94f63cc1232b35fa4aed298a8ed9e4ffad8 ?

rooftopcellist committed:
Merge pull request #16 from elyezer/update-subs

Update awxkit subscriptions page object

shanemcd committed:
Use forked rhsm

Just to be safe

shanemcd committed:
Add rhsm source, fix tests

shanemcd committed:
Mock tests that talk to database

rooftopcellist committed:
Subscriptions: Add customer JSON parser for Certificate

rooftopcellist committed:
Subscriptions: add migration to clear old license setting

rooftopcellist committed:
Subscriptions: Properly consume full quantity of chosen subscription

rooftopcellist committed:
Unattach unnecessary pools when refreshing_entitlement
  * fixed flake8

shanemcd committed:
Fix JS lint

shanemcd commented:
recheck

rooftopcellist committed:
init removal of unneeded license code

rooftopcellist committed:
Subscriptions: support satellite login flow

rooftopcellist committed:
Subscriptions: Add Login flow functionality for Satellite
  * Add Satellite boolean to Subscription config
  * Add explicit perm via ACL to subman config
  * Format Sat API data to work with existing parsing logic

rooftopcellist committed:
Subscriptions: Parse rhsm.conf directly, without rhsm library
  * Remove content check
